### PR TITLE
feat(users): Refuse a second account per email at signup without revealing it

### LINKED ISSUE #7854: Refuse a second account per email address at signup, without revealing that an account exists
Part of [#7437](https://github.com/freelawproject/courtlistener/issues/7437). **This stops the bleeding** — after this, the duplicate population stops growing and the backfill becomes a finite job.

## Context

CourtListener currently allows any number of accounts under one email address. That's the root cause of the confusion in [#7437](https://github.com/freelawproject/courtlistener/issues/7437): people create a second account, find none of their alerts, and think their data is gone. One member ended up paying for two memberships.

Requirement 1 of that issue is that **it should not be possible to determine whether an address has a CourtListener account.** Blocking duplicate registration naïvely breaks that: "this email is already in use" is a perfect existence oracle.

The codebase already has the right pattern in two places. `CustomPasswordResetForm.save` in `cl/users/forms.py` and `request_email_confirmation` in `cl/users/views.py` both render the success page unconditionally and then send mail — either the real thing, or `emails["no_account_found"]` from `cl/users/utils.py` saying no account exists here. The signal moves from the HTTP response, which anyone can see, to the inbox, which only the address owner can read. Mirror that.

### 1. The check

Add a uniqueness check to `UserCreationFormExtended` in `cl/users/forms.py`, querying with Django's `Lower("email")` — **not** Python's `str.lower()`. The two disagree on some non-ASCII input and the collation is locale-dependent, so folding in Python while the database index folds in SQL means one direction lets a duplicate past the form and fails at the index later, and the other rejects a legitimate address. Email local parts are not ASCII-restricted the way usernames are.

Exclude `self.instance.pk` from the query. That's also what makes the stub path below work: when the form is bound to an existing stub for that address, the only match is the instance itself.

### 2. Never reveal the result

- Always redirect to `register_success`, byte-identical to a successful registration.
- When the address is taken, create nothing and send mail to it: an account already exists, it can be reached by signing in **with this email address**, and here's a password reset link. Add it alongside the others in `cl/users/utils.py`.

Shipping after the email-login issue is what lets that email say something genuinely useful instead of "you have an account but we won't tell you the username."

### 3. Reject usernames that match another account's email address

Because `ASCIIUsernameValidator` permits `@` and `.`, `victim@example.com` is a registerable *username* today, which interferes with email login. Reject new usernames that equal another account's email address — reporting **only that the username is taken**, never why. Same reason as above: the reason is an existence oracle.

The census issue produces an audit list of existing collisions; this issue only needs to stop new ones.

## Gotchas

> [!IMPORTANT]
> **Put the check on the registration form only — not on `UserForm`.**
>
> `view_settings` validates `UserForm`, which also has an `email` field, and both forms share `CleanEmailMixin`. A check placed in the mixin would stop **every user who currently has a duplicate from saving their settings page**, because the address they legitimately own is "taken" by their other account. Excluding `self.instance.pk` does not save you — it's a *third* account holding the address.
>
> `UserForm` gets the same check in the last issue of the sequence, after the backfill has removed the duplicates. Please add a regression test that an existing duplicate can still save their profile.

**Keep the stub takeover for now.** `register` looks for an existing stub account on the submitted address and binds the form to it, which is how people who first appear via a Neon donation claim a real account. That branch must run before the uniqueness check and must stay non-enumerating.

Note that [#7785](https://github.com/freelawproject/courtlistener/issues/7785) plans to remove stub accounts entirely, which will delete this branch and the takeover vector it carries. **These two issues are not blocked on each other**, but they touch the same function, so whoever lands second should expect to adapt. If [#7785](https://github.com/freelawproject/courtlistener/issues/7785) has already landed, this issue gets simpler: no stub branch, no instance-exclusion subtlety.

**Races are acceptable.** hCaptcha, the honeypot (`check_honeypot(field_name="skip_me_if_alive")`) and `ratelimiter_unsafe_10_per_m` mean the only way past a form-level check is two genuinely simultaneous signups on one address. The backfill sweeps up whatever slips, and the unique index at the end of the sequence closes the window for good. Don't over-engineer locking here.

**Response timing still leaks.** Matching response bodies is not sufficient. The taken-address path sends one email and creates nothing; the fresh path creates a `User`, a `UserProfile` and a `Token`, runs the recursive `generate_recap_email` uniqueness loop, and sends *two* emails — one of them to `MANAGERS` — all synchronously via `send_mail` inside the view. The difference is measurable by anyone with a stopwatch. Route both paths' mail through the same asynchronous mechanism so the response *shape* matches, not just its text. This is a pre-existing property of `register`, but this is the issue that makes the promise, so the fix belongs here.

## A note on scope

Requirement 1 is achievable for email addresses. It is **not** achievable for usernames, and the issue is worth restating that way: usernames appear in public tag and prayer URLs (`/tags/<username>/`, `/prayers/<username>/`), and the registration form still reports username collisions. Don't spend effort chasing username non-enumerability.

## Dependencies

- **Blocked by:** the `LOWER(email)` index issue. Should ship after the email-login issue so the notification email can point at it.
- **Related:** [#7785](https://github.com/freelawproject/courtlistener/issues/7785) — see above.

## Done when

- [ ] Registering with a taken address produces the same response as registering with a fresh one, including timing shape.
- [ ] The address owner receives a useful email naming the sign-in options.
- [ ] No account is created for the duplicate attempt.
- [ ] A username equal to another account's email is rejected as "taken", with no explanation of why.
- [ ] **An existing duplicate user can still save their settings page** (regression test).
- [ ] The stub-account takeover path still works.

### PR DESCRIPTION
**Pull request by mlissner**
## Fixes

Fixes: #7854 (part of #7437); Stacked on #7900.

## Summary
This PR finally stops new duplicate accounts from being created under one email address. I do so while maintaining our guarantee not to leak whether somebody is a CL user. 

In future PRs, I'll clean up the dups we already have.

**The check** (`UserCreationFormExtended`, `cl/users/forms.py`)
- `clean_email` runs the existing blocklist check, then looks for other accounts holding the address. 

 - `clean_username` now rejects any new username that contains an at-sign. At first, we allowed email addresses as usernames, as we do now, but if we do that, we have to ensure that nobody registers a username that's the same as an existing user's email. If they were to do that, they could block the existing person from logging into their account. Not great. 

    Instead, we block new accounts containing `@`, which fixes the whole issue.

 - If somebody tries to create an account using somebody else's email, that's fine. We already handle that by selling the creator that the account was created successfully, and telling the other user by email that somebody attempted to create an account with the email address. (Usually this is actually somebody who forgot they already have an account, so in that case, the email also has their log in info, so they can log in.)

 - To prevent timing attacks, we use celery tasks to send emails and we do a password hash when pretending to create an account.

## Documentation

I don't think users need anything documented about this. They simply can't create secondary accounts, nor use their email as a username, and that's fine.

## Deployment

**This PR should:**
- [ ] `skip-deploy` (skips everything below)
    - [ ] `skip-web-deploy`
    - [ ] `skip-celery-deploy`
    - [x] `skip-cronjob-deploy`
    - [x] `skip-daemon-deploy`

Web and Celery both need to deploy together: the view enqueues two new tasks that the workers must know about.

## AI Disclosure
- [x] Parts of this PR were created with the help of an AI tool, and I have carefully reviewed all of its content and take full responsibility for it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01JTZfbnoKqCNrrvsqPk1iEP

### CODE REVIEW

**Comment by mlissner** [comment]
Re the Semgrep `django-no-csrf-token` finding on `cl/users/templates/register/login.html`:

- The sign-in form already carries the token. The `<form>` tag line reads `<form action="{% url "sign-in" %}" method="post">{% csrf_token %}`, so this is a false positive of the rule, likely because the tag sits on the same line as the opening `<form>`.
- That file is not part of this PR's diff. It changed in #7900 (a label text edit on line 29, not the form tag), and Semgrep surfaced it here because it compares against `main` rather than this PR's stacked base branch.

No change made here for it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cl/users/views.py [hunk 1]
```diff
@@ -544,6 +549,21 @@ def register(request: HttpRequest) -> HttpResponse:
             consent_form = OptInConsentForm(request.POST)
             if form.is_valid() and consent_form.is_valid():
                 cd = form.cleaned_data
+                # Every successful-looking outcome below, including a refused
+                # duplicate, must produce this exact redirect. Anything that
+                # differs between the outcomes tells an observer whether the
+                # address already had an account.
+                get_str = f"?next={urlencode(redirect_to)}&email={urlencode(cd['email'])}"
+                success_url = reverse("register_success") + get_str
```
Comment (mlissner) [inline cl/users/views.py]: This can go at the top under `redirect_to` or can get refactored into it. It's weird they're so far apart.

File: cl/users/forms.py [hunk 2]
```diff
@@ -195,6 +237,40 @@ class Meta(UserCreationForm.Meta):
             "username": forms.CharField,
         }
 
+    def _other_accounts_using_email(self, value: str) -> QuerySet[User]:
```
Comment (mlissner) [inline cl/users/forms.py]: This should be renamed and just return a bool, so that the `exists()` method isn't needed below.

File: cl/users/tests.py [hunk 3]
```diff
@@ -4384,6 +4384,267 @@ async def test_claiming_a_stub_account_issues_a_fresh_unguessable_key(
...
+        await sync_to_async(UserProfileWithParentsFactory.create)(
+            user__username="ada", user__email="ada@example.com"
+        )
+
+        errors_seen = []
+        for username in [
+            "ada@example.com",  # Another account's address
+            "Ada@Example.com",  # Same, different case
+            "nobody@example.com",  # An address with no account
+            "mal@ory",  # Not even an address, just an "@"
```
Comment (mlissner) [inline cl/users/tests.py]: This is fine. Let's allow this and make it match emails a little more exactly. We have a regex for email-like things. Use it so that `mal@ory` is ok, but things like `mal@or.y` are not.

**Review by mlissner — COMMENTED** [review]
A few comments, please.

File: cl/users/forms.py (hunk 2 above)
Comment (mlissner) [reply cl/users/forms.py]: Done in e038ef07: renamed to `_email_is_taken`, returns a bool, and `clean_email` just assigns it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cl/users/tests.py (hunk 3 above)
Comment (mlissner) [reply cl/users/tests.py]: Done in e038ef07. The validator now uses `LOOKS_LIKE_EMAIL` from `cl.lib.AuthenticationBackend`, the same test the sign-in backend applies, so anything it lets through can't be mistaken for an address at sign-in. `mal@ory` registers fine (new test), while `mal@or.y` and real addresses are refused with "Usernames cannot be email addresses." The PR body's "contains an at-sign" wording is now slightly off; I left it since you wrote it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by mlissner** [comment]
@claude review once

File: cl/users/views.py [hunk 4]
```diff
@@ -544,6 +559,14 @@ def register(request: HttpRequest) -> HttpResponse:
             consent_form = OptInConsentForm(request.POST)
             if form.is_valid() and consent_form.is_valid():
                 cd = form.cleaned_data
+                if form.email_taken:
+                    # Somebody already holds this address. Create nothing, and
+                    # tell only the address owner, by email, which nobody else
+                    # can read. Mail goes through the same async path as the
+                    # real signup so the response takes the same shape.
+                    notify_existing_account_holder.delay(cd["email"])
+                    return HttpResponseRedirect(success_url)
```
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/users/views.py:580]: Agreed, and fixed in 1de4e64a. The refused branch now runs `User().set_password(cd["password1"])` before enqueueing, the same throwaway-hash idiom `EmailOrUsernameModelBackend` uses, so both paths pay for one password hash. A test patches `set_password` and asserts the refused path calls it exactly once.

What remains different is the fresh path's handful of small inserts (User, UserProfile, Token, the recap-email uniqueness query). Those are a few milliseconds against a hash in the tens to hundreds, and the linked issue explicitly accepts that residual rather than moving account creation off the request path.

The review body's extra finding about the stub lookup using the unstripped `request.POST["email"]` is also fixed in the same commit: the lookup now uses the stripped address the form cleans, with a test that `"  donor@example.com  "` still claims the stub.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cl/users/forms.py [hunk 5]
```diff
@@ -195,6 +237,40 @@ class Meta(UserCreationForm.Meta):
...
+
+    def clean_email(self) -> str:
+        """Run the shared email checks, then record whether another account
+        already holds this address in ``email_taken``.
+
+        A taken address is deliberately not a validation error. See the class
+        docstring for why.
+        """
+        email = super().clean_email()
+        self.email_taken = self._other_accounts_using_email(email).exists()
```
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/users/forms.py]: This race is real but deliberately out of scope here. Issue #7854 says so directly: hCaptcha, the honeypot, and the per-minute rate limit mean the only way past a form-level check is two genuinely simultaneous signups on one address, the backfill sweeps up whatever slips, and the `LOWER(email)` unique index at the end of the sequence closes the window for good, so don't add locking in this PR. Not changing it here; the constraint lands in its own issue.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (mlissner) [reply cl/users/forms.py]: This will come later, in https://github.com/freelawproject/courtlistener/issues/7860

**Comment by mlissner** [comment]
And one more from me for triage/review, Alberto. Trying to keep my pace up and finally resolve this. 

This PR finally blocks multiple accounts/email via application code. Now I shift to cleanup.

**Review by nikollycardoso — APPROVED** [review]
This looks great!

Just an observation, I found an edge case worth flagging. I was able to register the same email twice through the UI with two separate tabs clicking "Register" at nearly the same time:

pk 2794  test   test22@email.com  04:30:31.75
pk 2795  test2  test22@email.com  04:30:32.19

I don't think it should block this, but maybe we can add a unique index when the duplications are clean.

**Comment by mlissner** [comment]
Thanks for the review. I'm impressed you pulled that off. I'd guess it would be hard to time correctly. 

The unique index is planned in https://github.com/freelawproject/courtlistener/issues/7860. 

This good to go otherwise?

**Comment by nikollycardoso** [comment]
Thanks! It wasn't hard to time at all, just two browser tabs side by side haha. Yes, good to go otherwise :+1:
